### PR TITLE
CMR-6853: Toggle-ability of CWIC or CMR Granule OSDDs

### PR DESCRIPTION
**Pull request by mitchstartzel**
* Added Flipper gem to allow toggling of which OSDD server to use at deployment time
* Updated OSDD-related tests to use the correct toggle at test time
* Added new tests to verify OSDD insertion behavior for when CWIC is enables

### CODE REVIEW

File: app/models/collection.rb [hunk 1]
```diff
@@ -124,13 +124,20 @@ def to_open_search_collections(document, hits, params, url, for_atom = true, cwi
...
-          add_link_as_child(doc, node, "#{ENV['opensearch_url']}/granules.atom?clientId=#{params[:clientId]}&shortName=#{short_name}&versionId=#{version_id}&dataCenter=#{data_center}", 'application/atom+xml', NEW_REL_MAPPING[:search], 'Search for granules')
-          add_link_as_child(doc, node, "#{ENV['opensearch_url']}/granules/descriptor_document.xml?collectionConceptId=#{id}&clientId=#{params[:clientId]}", 'application/opensearchdescription+xml', NEW_REL_MAPPING[:search], link_title)
+        else
+          if has_granules
+            add_link_as_child(doc, node, "#{ENV['opensearch_url']}/granules.atom?clientId=#{params[:clientId]}&shortName=#{short_name}&versionId=#{version_id}&dataCenter=#{data_center}", 'application/atom+xml', NEW_REL_MAPPING[:search], 'Search for granules')
+          end
+          if create_cwic_osdd_link || has_granules
+            link_title = "Granule OpenSearch Descriptor Document"
+            if Flipper.enabled?(:use_cwic_server)
+              Rails.logger.info "Flipper says use cwic server!"
```
Comment (jceaser) [inline app/models/collection.rb:134]: is this debug or info? Looks like DEBUG info. Why does anyone care to see this for every call?

File: Gemfile [hunk 2]
```diff
@@ -15,6 +15,7 @@ gem 'loofah', '>= 2.3.1'
 gem 'rack', '~> 2.0'
 gem 'ffi', '1.9.24'
 gem 'rails-controller-testing'
+gem 'flipper'
```
Comment (jaybarra) [inline Gemfile:18]: Does rubocop have similar 'in alphabetical order' rules for requires?

Comment (jceaser) [inline Gemfile:18]: why is a gem needed to read envs?

**Review by jaybarra — APPROVED** [review]

File: spec/views/collections/opensearch_collections_tagging_spec.rb [hunk 3]
```diff
@@ -22,29 +22,25 @@ def app
...
+        entry_osdd_link = entry.at_xpath('os:link[@title = \'Granule OpenSearch Descriptor Document\']', 'os' => 'http://www.w3.org/2005/Atom')
         entry_tag_value = entry.at_xpath('echo:tag/echo:tagKey', 'os' => 'http://www.w3.org/2005/Atom', 'echo' => 'https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html#atom').text
 
         # TEST entries
         if([2,5,8].include?(index+1))
           expect(entry_tag_value).to eq('org.ceos.wgiss.cwic.granules.test')
           # test will NOT have the CWIC OSDD link when the header is NOT present
-          expect(!entry_cwic_link.nil?).to be true
-          expect(entry_cmr_link.nil?).to be true
+          expect(!entry_osdd_link.nil?).to be true
```
Comment (jceaser) [inline spec/views/collections/opensearch_collections_tagging_spec.rb:32]: could you say the following:

expect(entry_osdd_link.nil?).to be false

File: spec/views/collections/cwic_enabled_opensearch_collections_tagging_spec.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+describe 'various provider granule OpenSearch API search behavior'  do
+  include Rack::Test::Methods
+
+  def app
+    Rails.application
+  end
+
+  before (:each) do
+    Flipper.enable(:use_cwic_server)
```
Comment (jceaser) [inline spec/views/collections/cwic_enabled_opensearch_collections_tagging_spec.rb:11]: so maybe this is the reason you have a library, you can simulate settings for tests?

Comment (mitchstartzel) [reply spec/views/collections/cwic_enabled_opensearch_collections_tagging_spec.rb:11]: Yes, also a valid reason. Another is that Flipper allows us to be more performant by only checking the value of the ENV var once, and then using the result of that check as long as it's deployed.

File: Gemfile (hunk 2 above)
Comment (mitchstartzel) [reply Gemfile:18]: No rules about require order, that I could find.

File: app/models/collection.rb (hunk 1 above)
Comment (mitchstartzel) [reply app/models/collection.rb:134]: Agreed - was useful for test deployments for searching Splunk, but now no longer necessary.

Comment (jceaser) [reply app/models/collection.rb:134]: then you will remove it?

Comment (mitchstartzel) [reply app/models/collection.rb]: Removed. 👍

File: spec/views/collections/opensearch_collections_tagging_spec.rb (hunk 3 above)
Comment (mitchstartzel) [reply spec/views/collections/opensearch_collections_tagging_spec.rb]: Fixed this in multiple files.

**Review by jceaser — APPROVED** [review]

**Pull request merged**